Web UI: Integrations Enroll redesign

alexhemard opened the pull request:
## Description
Resolves: https://github.com/gravitational/teleport/issues/57612
`e` PR: https://github.com/gravitational/teleport.e/pull/7144

Implements new Integration Enrollment [design](https://www.figma.com/design/uLevdNsEnIvvLDSZ9sQqXM/Discover-Access?node-id=3832-23143&t=TsBgJ3nFvdQmFwrz-4) 

- New tile design
- Adds Sorting, Filtering by search string, tags
- 'Add New Bot' redirects to Integration page with Bots filter applied
- New [AWS Icons](https://github.com/gravitational/teleport/issues/57612#issuecomment-3168210007)

## Demo

https://github.com/user-attachments/assets/cc560cd3-dcac-4cce-9f6c-8782371a6fe1

## Notes

Will need a followup PR to clean up 'old design' Components after `e` update is merged

Changelog: Updated Enroll Integration page design

alexhemard committed:
feat(Integrations): init redesign

alexhemard committed:
mv to Integrations/Enroll/Shared

alexhemard committed:
add missing license

alexhemard committed:
feedback, search tile descriptions

kimlisa commented:
haven't looked at the code yet, i just did some QA testing:

1. In community edition the filter types for `device trust, notification, and scim` gives me empty result. These are enterprise features. Do we want to render these tiles and have `requires upgrading to enterprise` CTA like others? Or should we not render these types in the dropdown? I'd lean towards rendering with CTA to stay consistent... but i'd double check with UX folks

2. When i click on filters, my URL includes `pinnedOnly=false` which doesn't make sense since we don't support pinning

3. A `guided` plugin tile once enrolled converts to `self-hosted` tile. We have like 4 different types of tiles: disabled requires upgrade or RBAC, self-hosted (click out to guide), guided + once enrolled prevent clicking tile to see guide (eg: slack), guided + once enroll retain click behavior (eg: okta)

4. Stories inside `teleport > integrations > enroll > picker` and `teleporte > integrations > picker` are broken

5. If you enter a invalid `kind` in the URL, it is silently ignored with no results (at least in the unified resources page it errors out saying what's wrong b/c there is a validation in the backend). in [session recording](https://localhost:3000/web/cluster/kimlisa35.cloud.gravitational.io/recordings) page, invalid types are just removed from the URL since it's client filtering (i'd lean towards this behavior)

6. if you hover over a disabled tile, it loses its border:
<img width="641" height="328" alt="image" src="https://github.com/user-attachments/assets/b3525d50-c42c-42bf-ac89-eee313ffcf6d" />

gzdunek reviewed (commented):
First pass.

alexhemard commented:
> When i click on filters, my URL includes pinnedOnly=false which doesn't make sense since we don't support pinning

@kimlisa This is leveraging the `useUrlFiltering` hook and it's appending the `isPinned` param. There's a [check to add the URL param if it's defined](https://github.com/gravitational/teleport/blob/master/web/packages/teleport/src/components/hooks/useUrlFiltering/encodeUrlQueryParams.ts#L52-L54), but there's a [default parameter of `false`](https://github.com/gravitational/teleport/blob/master/web/packages/teleport/src/components/hooks/useUrlFiltering/encodeUrlQueryParams.ts#L39), so always gets added. 

This also pops up in Access Lists search

<img width="746" height="375" alt="pinnedonly" src="https://github.com/user-attachments/assets/efef19b2-205e-4018-a2e4-cec2b327d068" />

Seems like removing the default `false` option would prevent this behavior or should I roll my own url query filtering if `useUrlFiltering` is not appropriate here 🤔

alexhemard commented on web/packages/teleport/src/Integrations/Enroll/Shared/FilterPanel.tsx:
```diff
@@ -0,0 +1,201 @@
+  );
+};
+
+const InputWrapper = styled.form`
```
It might be more pragmatic/consistent to scrap the Magnifier Icon from the design and use `DataTable.InputSearch` which would match the Enroll Resource search bar, but doesn't support adding an icon.

gzdunek replied:
Using `DataTable.InputSearch` sounds like a good option.

alexhemard committed:
fix storybook

alexhemard committed:
fix missing border for disabled tile hover

alexhemard committed:
replace useUrlFiltering, feedback

alexhemard commented:
Pushed changes:
- Replaced `DebouncedSearchInput` with `DataTable.Input`
- Replaced `useUrlFiltering` with invalid tag filtering 
- Enrolled plugin no longer displays as self-hosted
- misc feedback, cleanup

alexhemard commented on web/packages/teleport/src/Integrations/Enroll/Shared/FilterPanel.tsx:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    onSearch(debouncedTerm);
+  }, [debouncedTerm]);
```
Thinking it would if it's updating the URL parameters on change.
Replacing it with DataTable.SearchInput would only update on submit, so it wouldn't be necessary

alexhemard commented on web/packages/teleport/src/Integrations/Enroll/Shared/FilterPanel.tsx:
```diff
@@ -0,0 +1,201 @@
+      <StyledInput
+        placeholder={placeholder}
+        autoFocus
+        max={100}
```
Replaced with `DataTable.InputSearch`

alexhemard committed:
remove bot name filter hack

alexhemard commented on web/packages/teleport/src/Integrations/Enroll/Shared/state.ts:
```diff
@@ -0,0 +1,213 @@
+  );
+}
+
+export function useIntegrationPickerState(): [
```
Callout: This is mostly repurposed from https://github.com/gravitational/teleport/blob/master/web/packages/teleport/src/SessionRecordings/list/state.ts but this hook feels like it could be generic 🤔

gzdunek replied:
It might be a good idea to try this if it’s not too much work. But I'd do this outside this PR, since it's already quite large.

alexhemard committed:
simplify useIntegrationPickerState hook

gzdunek commented on web/packages/teleport/src/Integrations/Enroll/Shared/Tile.tsx:
```diff
@@ -0,0 +1,307 @@
+    ? { external: false, url: cfg.getIntegrationEnrollRoute(spec.kind) }
+    : null;
+
+  const hasAccess = (() => {
+    if (spec.kind === IntegrationKind.ExternalAuditStorage) {
+      const externalAuditStorageEnabled =
+        cfg.entitlements.ExternalAuditStorage.enabled;
+      return (
+        hasIntegrationAccess &&
+        hasExternalAuditStorage &&
+        externalAuditStorageEnabled
+      );
+    }
+
+    return hasIntegrationAccess;
+  })();
```
I'm not against IIFE, but it makes reading code a bit harder as indentation increases, which makes it more difficult to see where the function returns or ends.

WDYT of something like this:
```tsx
let hasAccess = hasIntegrationAccess;
if (spec.kind === IntegrationKind.ExternalAuditStorage) {
  hasAccess &&=
    cfg.entitlements.ExternalAuditStorage.enabled && hasExternalAuditStorage;
}
```

gzdunek commented on web/packages/teleport/src/Integrations/Enroll/Shared/Tile.tsx:
```diff
@@ -0,0 +1,307 @@
+    return hasIntegrationAccess;
+  })();
+
+  const renderBadge = (() => {
+    if (spec.kind === IntegrationKind.ExternalAuditStorage) {
+      const externalAuditStorageEnabled =
+        cfg.entitlements.ExternalAuditStorage.enabled;
+
+      return () => {
+        return renderExternalAuditStorageBadge(
+          hasExternalAuditStorage,
+          externalAuditStorageEnabled
+        );
+      };
+    }
+    if (!hasIntegrationAccess) {
+      return () => {
+        return <GenericNoPermBadge noAccess={!hasIntegrationAccess} />;
+      };
+    }
+
+    return undefined;
+  })();
```
This as well could be 
```suggestion
  const renderBadge = () => {
    if (spec.kind === IntegrationKind.ExternalAuditStorage) {
      const externalAuditStorageEnabled =
        cfg.entitlements.ExternalAuditStorage.enabled;

      return renderExternalAuditStorageBadge(
        hasExternalAuditStorage,
        externalAuditStorageEnabled
      );
    }
    if (!hasIntegrationAccess) {
      return <GenericNoPermBadge noAccess={!hasIntegrationAccess} />;
    }

    return undefined;
  };
```

gzdunek commented on web/packages/teleport/src/Integrations/Enroll/Shared/state.ts:
```diff
@@ -0,0 +1,188 @@
+  return urlParams.toString();
+}
+
+function sortArray(array: string[]) {
+  return array.toSorted((a, b) => a.localeCompare(b));
+}
+
+function arraysAreEqual(a: string[], b: string[]) {
+  if (a.length !== b.length) {
+    return false;
+  }
+
+  return JSON.stringify(sortArray(a)) === JSON.stringify(sortArray(b));
+}
+
+function filtersAreEqual(
+  a: IntegrationPickerFilters,
+  b: IntegrationPickerFilters
+) {
+  if (a.tags.length !== b.tags.length) {
+    return false;
+  }
+
+  return arraysAreEqual(a.tags, b.tags);
```
We don't seem to use these functions.

kimlisa commented on web/packages/teleport/src/Bots/Add/AddBots.tsx:
```diff
@@ -32,7 +31,10 @@ export function AddBots() {
           component={GitHubActionsFlow}
         />
-        <Route path={cfg.getBotsNewRoute()} component={AddBotsPicker} />
+        <Redirect
+          path={cfg.getBotsNewRoute()}
+          to={`${cfg.getIntegrationEnrollRoute()}?tags=bot`}
+        />
```
with this change, the `bot` nav section on click doesn't get highlighted like in the bottom image

<img width="402" height="295" alt="image" src="https://github.com/user-attachments/assets/edf2423c-6a20-4fc7-8aad-5751f9530fcd" />

before your changes:

<img width="372" height="305" alt="image" src="https://github.com/user-attachments/assets/0fd0fc6a-1d81-4d20-abb5-81cf90fe33ec" />

kimlisa commented on web/packages/teleport/src/Bots/Add/AddBotsPicker.tsx:
i think there are unused components and or functions in this file, can you double check?

i would put a `// TODO(alexhemard): delete in a follow up PR` to mark things for deletion later

kimlisa commented on web/packages/teleport/src/Integrations/Enroll/Shared/IntegrationPicker.test.tsx:
🙌

kimlisa reviewed (commented):
we're almost there

> This is leveraging the useUrlFiltering hook and it's appending the isPinned param. 

we make mistakes, this is one of those that got overlooked both by dev and code reviewers. i wouldn't rely on existing code and patterns to be always the best or accurate, i would always question it.

in fact, that access list `isPinned` got noticed by another dev who is working on something related and is in the process of fixing it. similarly if you notice anything odd (or notice any bug), ask in `#dev-frontend` or create an issue for it tagged with `ui`

alexhemard committed:
remove unused fns, tests

alexhemard committed:
renderBadge -> Badge

alexhemard committed:
feedback

alexhemard committed:
update RolesAnywhere

gzdunek commented on web/packages/teleport/src/Integrations/Enroll/Shared/FilterPanel.tsx:
```diff
@@ -0,0 +1,112 @@
+          placeholder="Search for integrations..."
+        />
+      </Box>
+      <Flex justifyContent="space-between" minWidth="419px">
```
This is rather a low level component, it probably shouldn't set the min width for the page.

gzdunek commented on web/packages/teleport/src/Integrations/Enroll/Shared/Tile.tsx:
```diff
@@ -0,0 +1,291 @@
+        (link && !link.external ? (
+          <BadgeGuided>Guided</BadgeGuided>
+        ) : (
+          <BadgeSelfHosted>
+            Self-Hosted
+            <NewTab size={14} ml={1} color="text.slightlyMuted" />
+          </BadgeSelfHosted>
```
The text and the icon should be aligned.

gzdunek commented on web/packages/teleport/src/Integrations/Enroll/Shared/FilterPanel.tsx:
```diff
@@ -0,0 +1,201 @@
+      }}
+    >
+      <Icons.Magnifier size={16} color="text.slightlyMuted" />
+      <StyledInput
```
What about this comment? :)

alexhemard committed:
feedback

alexhemard committed:
filter tags if no matching integrations

alexhemard committed:
update comments
